runtime/casts.cu

```diff
@@ -359,176 +359,211 @@ __device__ __inline__ bool __heq(const __bfloat a, const __bfloat b) {
   return (val != 0U) ? true : false;
 }
 
-// NOTE [ fp8 cast optimization ]
-//
-// For simplicity, we only provided fp8 <-> fp32 cast implementation, while
-// relying on any other fp cast in the form of target_fp <-> fp32 <-> fp8.
-// This avoids the complication of handling hardware specific instructions on
-// various compute capabilities.
-// But this simplicity could come at the cost of performance. In cuda_fp8.hpp,
-// 1. bf16 -> fp8 is done via bf16 -> float -> fp8
-// 2. fp16 -> fp8 is done with a conditional
-//    # if (> sm_89)
-//    fp16 -> fp8
-//    # else
-//    fp16 -> fp32 -> fp8
-//    # endif
-// 3. fp64 -> fp8 is handled explicitly as bitwise operations.
-// TODO consider cuda_fp8.hpp for performance optimized cast.
-__device__ __inline__ __e4m3 __float2e4m3(const float f) {
-  constexpr float f_const_zero = 0.f;
-  unsigned short _tmp_buffer;
-  __e4m3 val;
+template <int align>
+__device__ __inline__ Array<__e4m3, 2, align> __float2e4m3(
+    const Array<float, 2, align>& input) {
+  Array<__e4m3, 2, align> result;
+  uint16_t& result_scalar = *reinterpret_cast<uint16_t*>(&result);
   asm("{cvt.rn.satfinite.e4m3x2.f32 %0, %1, %2;}"
-      : "=h"(_tmp_buffer)
-      : "f"(f_const_zero), "f"(f));
-  memcpy(&val, &_tmp_buffer, sizeof(uint8_t));
-
-  return val;
+      : "=h"(result_scalar)
+      : "f"(input[1]), "f"(input[0]));
```

Comment:
Just wanted to make sure this order is intentional rather than `input[0], input[1]`, right?

Reply:
Yes, I was originally using `input[0], input[1]` and tests were failing